### PR TITLE
Serialize and restore ParsedChatRequest in chat sessions

### LINKED ISSUE #16561: AI Persistence: Serialize and Restore ParsedRequest
### Description

Currently, the `ParsedRequest` data (which contains parsed agent references, file references, and other structured request metadata) is not serialized and restored. After restoration, sessions lose this structured information.

The [`message`](https://github.com/eclipse-theia/theia/blob/7946ca874b04e6249d883e9a586f193194df1365/packages/ai-chat/src/common/chat-model.ts#L268) field in `ChatRequestModel` holds a `ParsedChatRequest` but is not currently serialized in [`toSerializable()`](https://github.com/eclipse-theia/theia/blob/7946ca874b04e6249d883e9a586f193194df1365/packages/ai-chat/src/common/chat-model.ts#L1640-L1650).

### Current Behavior

When a session is restored, request text is preserved but the parsed structure is lost.

### Desired Behavior

- Serialize `ParsedRequest` along with request text in `MutableChatRequestModel.toSerializable()`
- Restore `ParsedRequest` when loading persisted sessions in the [constructor](https://github.com/eclipse-theia/theia/blob/7946ca874b04e6249d883e9a586f193194df1365/packages/ai-chat/src/common/chat-model.ts#L1469-L1491)
- Preserve parsed data as much as possible

### Technical Considerations

- Add `toSerializable()` method to `ParsedRequest` or relevant classes
- Add deserialization logic in `MutableChatRequestModel` constructor for `ParsedRequest` data

### Related

- Follow up from initial implementation #16486

### PR DESCRIPTION
**Pull request by lucas-koehler**
#### What it does

Fixes #16561

Implement serialization of ParsedChatRequest data to preserve structured information (agent references, variable references, tool requests) across session persistence and restoration.

Previously, only raw text was preserved after session restoration, causing loss of parsed structure data. This affected UI rendering and prevented proper display of agent names, variables, and tool references in restored chat sessions.

Main code changes:

- Add serialization interfaces for ParsedChatRequest components in chat-model-serialization.ts (SerializableParsedRequestPart, SerializableToolRequest, SerializableResolvedVariable, etc.)
- Implement toSerializable() methods on ParsedChatRequestTextPart, ParsedChatRequestVariablePart, ParsedChatRequestFunctionPart, and ParsedChatRequestAgentPart classes
- Update MutableChatRequestModel to serialize/deserialize parsedRequest data via toSerializable() and deserializeParsedRequest() methods
- Serialize tool functions as id only and restore from the registry. For this, MutableChatModel and MutableChatRequestModel now require the tool registry as constructor parameter
- Add comprehensive test coverage for serialization/deserialization of all part types, including edge cases and complex mixed requests

[16561-pr.webm](https://github.com/user-attachments/assets/5b95201f-f6a9-420d-9cfa-5d290fec321f)

#### How to test

- Use the AI chat and chat using an agent and variable reference
- Restart theia
- Restore the session
- Observe: The agent, variable and tool function are now highlighted and provide additional information on hover

~**NOTE:** While the serialization also supports functions/tool requests, highlighting them does not work yet in general. See #16722~

#### Breaking changes

- [ ] This PR introduces breaking changes and requires careful review. If yes, the breaking changes section in the [changelog](https://github.com/eclipse-theia/theia/blob/master/CHANGELOG.md) has been updated.

#### Review checklist

- [x] As an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)
- [x] User-facing text is internationalized using the `nls` service (for details, please see the [Internationalization/Localization section](https://github.com/theia-ide/theia/blob/master/doc/coding-guidelines.md#internationalizationlocalization) in the Coding Guidelines)

#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

**Comment by lucas-koehler** [comment]
@sdirix That is a very good point! Thanks for catching that. I refactored the code accordingly to restore the tool requests from the registry in the chat service.
In the current refactoring, the chat service uses the registry directly  as the deserializer registry is currently only meant for response content but tool requests are part of the request.
Is this okay with you or do you think we should extend the registry now to also support request content?

**Comment by sdirix** [comment]
I will review at the latest next week.

File: packages/ai-chat/src/common/chat-model.ts [hunk 1]
```diff
@@ -1522,32 +1542,144 @@ export class MutableChatRequestModel implements ChatRequestModel, EditableChatRe
...
+                        partData.range,
+                        partData.variableName,
+                        partData.variableArg
+                    );
+                    if (partData.variableValue !== undefined) {
+                        varPart.resolution = {
+                            variable: {
+                                id: partData.variableId ?? 'restored',
+                                name: partData.variableName,
+                                description: 'Restored from serialized session'
```
Comment (sdirix) [inline packages/ai-chat/src/common/chat-model.ts]: Should we also store/restore the description?

File: packages/ai-chat/src/common/chat-model.ts [hunk 2]
```diff
@@ -1522,32 +1542,144 @@ export class MutableChatRequestModel implements ChatRequestModel, EditableChatRe
...
+                case 'var': {
+                    const varPart = new ParsedChatRequestVariablePart(
+                        partData.range,
+                        partData.variableName,
+                        partData.variableArg
+                    );
+                    if (partData.variableValue !== undefined) {
+                        varPart.resolution = {
+                            variable: {
+                                id: partData.variableId ?? 'restored',
```
Comment (sdirix) [inline packages/ai-chat/src/common/chat-model.ts]: I think if there is no id we should not restore it? And during storing we should make sure they have an id?

File: packages/ai-chat/src/common/chat-model.ts [hunk 3]
```diff
@@ -1522,32 +1542,144 @@ export class MutableChatRequestModel implements ChatRequestModel, EditableChatRe
...
+                                description: 'Restored from serialized session'
+                            },
+                            arg: partData.variableArg,
+                            value: partData.variableValue
+                        };
+                    }
+                    return varPart;
+                }
+                case 'function':
+                    // Create placeholder - will be restored by ChatService
```
Comment (sdirix) [inline packages/ai-chat/src/common/chat-model.ts:1615]: We're now using a placeholder approach. I wonder whether we should just construct the whole `ParsedChatRequest` in the service so we could skip the placeholder and create the "real" ones immediately. What do you think? Or do you prefer the placeholders?

File: packages/ai-chat/src/common/chat-model.ts (hunk 1 above)
Comment (lucas-koehler) [reply packages/ai-chat/src/common/chat-model.ts]: Yes, this can be quite easily done. I'll add it

File: packages/ai-chat/src/common/chat-model.ts (hunk 3 above)
Comment (lucas-koehler) [reply packages/ai-chat/src/common/chat-model.ts:1615]: I'd use the placeholder approach as this would result in a bigger refactoring and we use the pattern of restoring stuff (e.g. changesets) in the service already. Thus, I'd not like to do it as part of this PR but would be open to this being refactored in general as a separate task.

File: packages/ai-chat/src/common/chat-model.ts (hunk 2 above)
Comment (lucas-koehler) [reply packages/ai-chat/src/common/chat-model.ts]: I think the problem lies earlier: variable ids are only available if the variable is resolved during serialization (see packages/ai-chat/src/common/parsed-chat-request.ts). Currently we serialize the variable in any case.
We could think about serializing unresolved variables as text parts but there could be value in knowing this is a variable even if it was not/could not be resolved at the time of serialization.

Comment (lucas-koehler) [reply packages/ai-chat/src/common/chat-model.ts]: For now, I moved adding a fallback value of the id to the serialization function.

**Comment by lucas-koehler** [comment]
@sdirix I added serialization of the description. I also removed the 2 changelog commits during rebase: They cancelled each other and caused conflicts.
I also moved the initialization of the variable id fallback value to the serialization function of a variable part.
For now I left serializing and deserializing variable parts without an id. IMO the core question is whether we want to serialize variable parts with an unresolved variable at all. IMO this can be a follow up issue. WDYT?
